File: backend/app/schemas/article.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.tag import TagResponse
from app.schemas.user import UserResponse


class ArticleCreate(BaseModel):
    """Request body for creating an article."""

    title: str = Field(..., min_length=1, max_length=255)
    content: str = Field(..., min_length=1)
    summary: Optional[str] = None
    cover_image_url: Optional[str] = None
    tag_ids: Optional[list[uuid.UUID]] = None

    model_config = ConfigDict(from_attributes=True)


class ArticleUpdate(BaseModel):
    """Request body for updating an article."""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    content: Optional[str] = Field(None, min_length=1)
    summary: Optional[str] = None
    cover_image_url: Optional[str] = None
    is_published: Optional[bool] = None

    model_config = ConfigDict(from_attributes=True)


class ArticleResponse(BaseModel):
    """Full article representation including content."""

    id: uuid.UUID
    title: str
    slug: str
    content: str
    summary: Optional[str] = None
    cover_image_url: Optional[str] = None
    is_published: bool
    view_count: int
    upvote_count: int
    downvote_count: int
    comment_count: int
    tags: list[TagResponse] = []
    author: UserResponse
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ArticleListResponse(BaseModel):
    """Article summary for list views (no content field)."""

    id: uuid.UUID
    title: str
    slug: str
    summary: Optional[str] = None
    cover_image_url: Optional[str] = None
    is_published: bool
    view_count: int
    upvote_count: int
    downvote_count: int
    comment_count: int
    tags: list[TagResponse] = []
    author: UserResponse
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/article.ts
import { PublicUser } from "./user";

export interface Article {
  id: string;
  title: string;
  slug: string;
  content: string;
  summary: string;
  cover_image_url: string | null;
  author: PublicUser;
  tags: string[];
  upvotes: number;
  downvotes: number;
  view_count: number;
  comment_count: number;
  is_published: boolean;
  is_featured: boolean;
  user_vote: "up" | "down" | null;
  is_bookmarked: boolean;
  created_at: string;
  updated_at: string;
}

export interface ArticleCreate {
  title: string;
  content: string;
  summary: string;
  cover_image_url?: string;
  tags?: string[];
  is_published?: boolean;
}

export interface ArticleUpdate {
  title?: string;
  content?: string;
  summary?: string;
  cover_image_url?: string;
  tags?: string[];
  is_published?: boolean;
}

export interface ArticleListItem {
  id: string;
  title: string;
  slug: string;
  summary: string;
  cover_image_url: string | null;
  author: PublicUser;
  tags: string[];
  upvotes: number;
  downvotes: number;
  view_count: number;
  comment_count: number;
  is_published: boolean;
  is_featured: boolean;
  user_vote: "up" | "down" | null;
  is_bookmarked: boolean;
  created_at: string;
}

export interface ArticleVote {
  article_id: string;
  vote: "up" | "down";
}
